File: Sep/Day10/minimum-number-of-people-to-teach.py
#  Time: O(M * N)  ::  Space: O(M + N)
class Solution:
    def minimumTeachings(self, n: int, languages: List[List[int]], friendships: List[List[int]]) -> int:
        intersectionOfLang = set()
        
        for friendship in friendships:
            langMap = {}
            isCommon = False
            for lang in languages[friendship[0]-1]:
                langMap[lang] = 1
            for lang in languages[friendship[1]-1]:
                if lang in langMap:
                    isCommon = True
                    break
            if not isCommon:
                intersectionOfLang.add(friendship[0]-1)
                intersectionOfLang.add(friendship[1]-1)
        
        maxPeopleKnownLang = 0
        countOfPeopleKnownLang = [0]*(n+1)

        for friendship in intersectionOfLang:
            for lang in languages[friendship]:
                countOfPeopleKnownLang[lang]+=1
                maxPeopleKnownLang = max(countOfPeopleKnownLang[lang], maxPeopleKnownLang)
        
        return len(intersectionOfLang) - maxPeopleKnownLang
        
        

File: Sep/Day10/minimum-number-of-people-to-teach.cc
//  Time: O(M * N)  ::  Space: O(M + N)
class Solution {
public:
    int minimumTeachings(int n, vector<vector<int>>& languages, vector<vector<int>>& friendships) {
        unordered_set<int> intersectionOfLang;

        for (auto& friendship : friendships) {
            int u = friendship[0] - 1, v = friendship[1] - 1;
            unordered_set<int> langMap(languages[u].begin(), languages[u].end());
            bool isCommon = false;
            for (int lang : languages[v]) {
                if (langMap.count(lang)) {
                    isCommon = true;
                    break;
                }
            }
            if (!isCommon) {
                intersectionOfLang.insert(u);
                intersectionOfLang.insert(v);
            }
        }

        vector<int> countOfPeopleKnownLang(n + 1, 0);
        int maxPeopleKnownLang = 0;

        for (int person : intersectionOfLang) {
            for (int lang : languages[person]) {
                countOfPeopleKnownLang[lang]++;
                maxPeopleKnownLang = max(maxPeopleKnownLang, countOfPeopleKnownLang[lang]);
            }
        }

        return (int)intersectionOfLang.size() - maxPeopleKnownLang;
    }
};
